checkdigit/crc.py

```diff
@@ -0,0 +1,102 @@
+This can be easily checked even by very simple circuitry,
+and can also be used to correct some errors.
+
+WARNING: THIS IS NOT A FAST IMPLEMENTATION OF CRC.
+
+If you want a fast implementation look elsewhere.
+
+
+"""
+from checkdigit._data import cleanse
+
+# WARNING: Data beginning with 0 must be as a string due to PEP 3127
+
+
+def calculate(data: str, polynomial: str, pad: str = "0") -> str:
+    """Adds a parity part onto the end of a block of data.
+
+    Args:
+        data: A string containing binary digits of any length
+        polynomial: A string of binary digits representing the polynomial being used
+        pad: "1" or "0" or nothing to be used to pad the data (defaults to 0)
+
+    Returns:
+        str: Check Value that should be appended to the stream
+    """
+    data = cleanse(data)
+    data += pad * (len(polynomial) - 1)
+    bitarray = list(data)
+    while len(bitarray) != len(polynomial) - 1:
+        for (bit, _) in enumerate(polynomial):
+            if polynomial[bit] == bitarray[bit]:
+                bitarray[bit] = "0"  # XOR calculation
+            else:
+                bitarray[bit] = "1"
+        while bitarray[0] == "0" and len(bitarray) >= len(polynomial):
+            bitarray.pop(0)
+
+    return "".join(bitarray)
+
+
+def validate(data: str, polynomial: str) -> bool:
+    """Validates whether the check digit matches a block of data.
+
+    Args:
+        data: A string containing binary digits including the check digit
+        polynomial: Polynomial to use
+
+    Returns:
+        bool: A boolean representing whether the data is valid or not
+    """
+    data = cleanse(data)
+    # the principle of CRCs is that when done again but with the check digit
+    # appended if the data is fine it should all be 0s
+    return "0" * (len(polynomial) - 1) == calculate(data, polynomial, "")
+
+
+def missing(data: str, polynomial: str) -> str:
+    """Calculates missing digits represented by a question mark.
+
+    Args:
+        data: A string containing a question mark representing a missing digit.
+        polynomial: What the polynomial that should be used is
+
+    Returns:
+        str: The missing binary digit
+    """
+    solutions = []
+    number = data.count("?")
+    if number == 0:
+        return "Invalid"  # if there are no ? to replace the algorithm will not work
+    permutations = 2 ** number  # number of different permutations that are possible
+    for permutation in range(permutations + 1):
+        tocheck = data
+        replacement = bin(permutation)[2:].zfill(
+            number
+        )  # gives us a nice binary number
+        for bit in replacement:
+            tocheck = tocheck.replace("?", bit, 1)  # only replaces one bit at a time
+        if validate(tocheck, polynomial):
+            solutions.append(replacement)
+            if len(solutions) == 2:
+                return "Invalid"
+    if len(solutions) == 1:
+        return solutions[0]
+    return "Invalid"
```

Comment:
Thanks @sapieninja. The workflows seem to be playing up, so I might close and reopen the PR to fix it. I ran the tests locally, and we're just missing a test for this line.

Would it also be possible to have another valid `crc.missing` test, just so I have a better understanding of the format? Thanks.

```
Name                     Stmts   Miss Branch BrPart  Cover   Missing
--------------------------------------------------------------------
checkdigit/__init__.py       1      0      0      0   100%
checkdigit/_data.py          7      0      2      0   100%
checkdigit/crc.py           35      1     20      1    96%   102
checkdigit/gs1.py           24      0      8      0   100%
checkdigit/isbn.py          26      0     11      0   100%
checkdigit/luhn.py          26      0     12      0   100%
checkdigit/parity.py        12      0      2      0   100%
checkdigit/upc.py            6      0      0      0   100%
--------------------------------------------------------------------
TOTAL                      137      1     55      1    99%
Coverage failure: total of 99 is less than fail-under=100
```

*(Ignore the RTD test failure. That's because I hadn't set up RTD when you opened the PR).* Thanks again for doing this @sapieninja :) I really appreciate the work you've done 👍